### PR TITLE
Prototype

### PR DESCRIPTION
**Pull request by richfitz**
This sketches out the approach for the odin to js compiler.  It might eventually merge back into the main odin package (I am leaning that way at present).

Despite the apparent size of this PR, this is only 287 lines of actual code (according to codecov).

* the javascript code in inst is generated out of https://github.com/mrc-ide/dopri-js
* there's a lot of lines in the README that are just generated too
* the tests are minimal at present, but will eventually converge on the main test suite used by odin (this is one reason why this code is probably destined for odin itself - it's not easily possible to write the test suite that can be repurposed easily)

Ideally it would be great to also get feedback on the generated javascript - the README.md file contains an example of a generated model that is pretty typical

### CODE REVIEW

File: inst/example/simple/index.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/dygraph/2.1.0/dygraph.min.css" />
+    <script src="odin.js"></script>
+  </head>
+  <body onload="odinInit()">
+    <p>
+      <b>time range: </b>
+      <input type="text" width="5" id="t0" value="0">
+      to
+      <input type="text" width="5" id="t1" value="50">
+    </p>
```
Comment (weshinsley) [inline inst/example/simple/index.html]: The input tags should end in /> probably - not that it really matters.

File: inst/example/simple/index.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    <p>
+      <b>parameters: </b>
+      <div id="odin_parameters">
+      </div>
+    </p>
```
Comment (weshinsley) [inline inst/example/simple/index.html]: More html nit-picking - I don't think a `<p>` tag can contain a `<div>` tag - perhaps replace the `<p>` with `<br/>&nbsp;<br/>` or some other spacing, if that's what they're here for, (and remove `</p>`)

(Again, this doesn't matter, but I was just looking at the HTML source with Firefox and dealing with its complaints...)

File: inst/models/sir.R [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+initial(I) <- I0
+initial(R) <- 0
+
+Births <- N / 75
+b <- 1 / 75
+N <- 1e7
+I0 <- user(1)
+beta <- user(24)
+sigma <- 12
+delta <- 1 / 5
```
Comment (weshinsley) [inline inst/models/sir.R:15]: The two examples above show the strength of this sort of modelling - beautifull minimal, clear, mathematical syntax, and `user()`. 

You could consider some bracketing on the deriv() functions - just for peasants like me who aren't very good at remembering BODMAS.

And I am disturbed by your brazen dividing of two integers (which gets the right answer in js...)

File: inst/dopri.js [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+var E7 = -1.0 / 40.0;
+// ---- DENSE OUTPUT OF SHAMPINE (1986)
+var D1 = -12715105075.0 / 11282082432.0;
+var D3 = 87487479700.0 / 32700410799.0;
+var D4 = -10690763975.0 / 1880347072.0;
+var D5 = 701980252875.0 / 199316789632.0;
+var D6 = -1453857185.0 / 822651844.0;
+var D7 = 69997945.0 / 29380423.0;
+// most of this can be made really quit private - everything not used
+// in the interface really.
```
Comment (weshinsley) [inline inst/dopri.js:245]: quit -> quite. Yep. I've even spotted a typo in your comments, such is the tremendous thoroughness of my review.

File: inst/dopri.js [hunk 5]
```diff
@@ -0,0 +1,574 @@
+(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(require,module,exports){
+(function (global){
+global.dopri = require('dopri');
+
```
Comment (weshinsley) [inline inst/dopri.js:4]: Yummy... that'll teach me to click the "Load Diff" button.

File: inst/dopri.js [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+        // known to work and (b) I work out how the delay lookup will
+        // work as that needs to feed into the derivative function in
+        // odd ways.
+        //
+        // TODO: need to protect us from walking off the end of the
+        // array (or starting within it).
+        //
+        // TODO: don't allow anything to happen with a zero-length
+        // history.
+        var it = h[0].length - 2;
```
Comment (weshinsley) [inline inst/dopri.js:478]: I'm going to assume that you're happy for this PR, with the state of the four "TODO" comments above?

**Review by weshinsley — CHANGES_REQUESTED** [review]
Not many ideas - it all looks good to me, and there were just a few tiny HTML compliance things, and a couple of other very minor comments...

File: inst/dopri.js (hunk 4 above)
Comment (richfitz) [reply inst/dopri.js:245]: A typo in a comment in the bundled upstream source no less - deleted here: https://github.com/mrc-ide/dopri-js/pull/4/files#diff-4349159aa3a8f26b0d335f7572d41d3eL45

File: inst/dopri.js (hunk 6 above)
Comment (richfitz) [reply inst/dopri.js:478]: again these are in the bundled library...

**Review by weshinsley — APPROVED** [review]

**Pull request merged**